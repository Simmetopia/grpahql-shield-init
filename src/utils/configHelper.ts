import { config } from 'dotenv';

config();

export const tokenSecret = (): string => {
  return process.env.TOKEN_KEY || '';
};
