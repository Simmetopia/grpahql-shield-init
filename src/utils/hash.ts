import { hash, verify } from 'argon2';

export const saltAndHash = (toHash: string) => {
  // argon2id = type:2
  return hash(toHash, { saltLength: 16, memoryCost: 2048, type: 2 });
};

export const compareHash = (hashToCompare: string, compareWith: string) => {
  return verify(hashToCompare, compareWith);
};
