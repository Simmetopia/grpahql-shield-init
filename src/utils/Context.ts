import { decode, verify } from 'jsonwebtoken';
import { tokenSecret } from './configHelper';

export interface Context {
  user?: any;
  verified: boolean;
}
export const createContext = async (request: any): Promise<Context> => {
  const token = request.headers.authorization || '';

  try {
    verify(extractTokenFromHeader(token), tokenSecret());
    /* Get user data from token or fetch from database  */
    const decodedToken = decode(token) || { sub: '' };
    return {
      verified: true,
    };
  } catch (error) {
    return {
      verified: false,
    };
  }
};

const extractTokenFromHeader = (authHeader: string) => {
  const headerString = authHeader.split(' ');
  if (headerString.length === 1) {
    return headerString[0];
  }
  return headerString[1];
};
