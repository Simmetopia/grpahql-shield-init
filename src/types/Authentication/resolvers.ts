import { ApolloError } from 'apollo-server';
import { GraphQLFieldResolver } from 'graphql';
import { sign } from 'jsonwebtoken';
import { toLower } from 'lodash';
import { tokenSecret } from '../../utils/configHelper';
import { Context } from '../../utils/Context';
import { compareHash, saltAndHash } from '../../utils/hash';

export interface AuthenticationInputArgs {
  email: string;
  password: string;
}

interface Authentication {
  authArgs: AuthenticationInputArgs;
}

export const login: GraphQLFieldResolver<any, Context, Authentication> = async (
  _,
  { authArgs: { email, password } },
  ctx,
) => {
  throw new Error('no implemented');
};

export const signup: GraphQLFieldResolver<any, Context, Authentication> = async (
  _,
  { authArgs: { email, password } },
  ctx,
) => {
  throw new Error('no implemented');
};
