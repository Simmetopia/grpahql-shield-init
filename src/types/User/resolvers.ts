import { ApolloError } from 'apollo-server';
import { GraphQLFieldResolver } from 'graphql';
import { Context } from '../../utils/Context';

interface FindUserInputArgs {
  id: string;
}

export interface PaginationInputArgs {
  take: number;
  skip: number;
}

interface GetUsers {
  pagination: PaginationInputArgs;
}

export const getUser: GraphQLFieldResolver<any, Context, FindUserInputArgs> = async (_, { id }, ctx) => {
  throw new Error('no implemented');
  expect(false).toBeTruthy();
};

export const getUsers: GraphQLFieldResolver<any, Context, GetUsers> = (_, args, ctx) => {
  throw new Error('no implemented');
  expect(false).toBeTruthy();
};

export const deleteUser: GraphQLFieldResolver<any, Context, FindUserInputArgs> = async (_, { id }, ctx) => {
  throw new Error('no implemented');
  expect(false).toBeTruthy();
};

export const getMe: GraphQLFieldResolver<any, Context, any> = (_, args, ctx) => {
  throw new Error('no implemented');
  expect(false).toBeTruthy();
};
