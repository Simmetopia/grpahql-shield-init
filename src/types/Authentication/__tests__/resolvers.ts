import * as jwt from 'jsonwebtoken';
import { Context } from '../../../utils/Context';
import { saltAndHash } from '../../../utils/hash';
import { login, signup } from '../resolvers';
jest.mock('jsonwebtoken');
jest.mock('../../../utils/Context');

describe('Authenticantion resolver tests', () => {
  beforeAll(async () => {
    /* Implement clean test setup here  */
  });

  beforeEach(() => {
    jest.resetAllMocks();
  });

  describe('Login', () => {
    test('When loggin in with correct credentials, should return auth token', async () => {
      throw new Error('no implemented');
      expect(false).toBeTruthy();
    });

    test('When loggin in with wrong credentials, should return auth token', async () => {
      throw new Error('no implemented');
      expect(false).toBeTruthy();
    });
    describe('Signup', () => {
      test('When signing up, if invited, should return token', async () => {
        throw new Error('no implemented');
        expect(false).toBeTruthy();
      });
      test('When signing up, if admin, should return token with admin priviledges', async () => {
        throw new Error('no implemented');
        expect(false).toBeTruthy();
      });
      test('When signing up, if user is invited, should return token with user priviledges', async () => {
        throw new Error('no implemented');
        expect(false).toBeTruthy();
      });

      test('When signing up, if user is invited, should delete invite', async () => {
        throw new Error('no implemented');
        expect(false).toBeTruthy();
      });
      test('When signing up, if not admin, and not invited, should throw error', async () => {
        throw new Error('no implemented');
        expect(false).toBeTruthy();
      });
    });
  });
});
