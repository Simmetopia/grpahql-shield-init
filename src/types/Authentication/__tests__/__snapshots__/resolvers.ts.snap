// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authenticantion resolver tests Login When loggin in with wrong credentials, should return auth token 1`] = `"Unknown login credentials"`;
